### ISSUE TITLE
"You need to call SQLitePCL.raw.SetProvider()" error when using Microsoft.EntityFrameworkCore.Sqlite 5.0.4 via C++\CLI wrapper

### ISSUE THREAD

**Issue by bairog** [dotnet/efcore#24493]
## File a bug
I have a solution with the following structure:

- DomainObjects - .NET 5.0 class library without any special dependencies, contains entities. Compiled as **x86**.
- DAL (references DomainObjects) - .NET 5.0 class library which references **Microsoft.EntityFrameworkCore.Sqlite 5.0.4** nuget package, contains DbContext and migrations (database is created at **d:\test.db**). Compiled as **x86**.
- DAL.Test (references DAL) - .NET 5.0 console test application. Compiled as **x86**.
- DAL.Interop (references DAL) - .NET 5.0 C++\CLI wrapper. Compiled  as **x86, Multi-byte Character Set, wchar_t-**.
- DAL.Interop.Test - C++ console test application. Compiled  as **x86, Multi-byte Character Set, wchar_t-**.

**DAL.Test** runs without any problems, but **DAL.Interop.Test** throws an exception (inside **DAL** class library):

> You need to call SQLitePCL.raw.SetProvider().  If you are using a bundle package, this is done by calling SQLitePCL.Batteries.Init().

**NB!** If I switch from **Microsoft.EntityFrameworkCore.Sqlite** provider to **Devart.Data.SQlite.EFCore** provider - both test applications (.NET and C++) run without any problems.

**NB2!** If I remember correctly code was working with no problems with **Microsoft.EntityFrameworkCore.Sqlite 3.1** on **.NET Core 3.1**.

### Include your code

Sample repo - https://github.com/bairog/EFCore5SQLiteCLRTest
Compile it (for some reason you need to compile twice for compiling **DAL.Interop.Test** project) and debug **DAL.Interop.Test.exe**

### Include stack traces

```
Source: SQLitePCLRaw.core
You need to call SQLitePCL.raw.SetProvider(). If you are using a bundle package, this is done by calling SQLitePCL.Batteries.Init().
   at SQLitePCL.raw.get_Provider()
   at SQLitePCL.raw.sqlite3_open_v2(utf8z filename, sqlite3& db, Int32 flags, utf8z vfs)
   at Microsoft.Data.Sqlite.SqliteConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnection(Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternal(Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open(Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Sqlite.Storage.Internal.SqliteDatabaseCreator.Exists()
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.Exists()
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.GetAppliedMigrations()
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.GetAppliedMigrations(DatabaseFacade databaseFacade)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.GetPendingMigrations(DatabaseFacade databaseFacade)
   at DAL.CRUD.CreateDbContext() in D:\EFCore5SQLiteCLRTest\DAL\CRUD.cs:line 14
```

### Include provider and version information
EF Core version: 5.0.4
Database provider: Microsoft.EntityFrameworkCore.Sqlite
Target framework: .NET 5.0
Operating system: Windows 10 x64 1909
IDE: Visual Studio 2019 16.9.2 REL Community

**Comment by bricelam** [comment]
Is `SQLitePCLRaw.batteries_v2.dll` getting trimmed/not getting copied to the output directory?

**Comment by bairog** [comment]
@bricelam `SQLitePCLRaw.batteries_v2.dll` is present in output directory. Don't know if it is trimmed - it is **6Kb**.

**Comment by bairog** [comment]
**.NET 6-preview5** was out almost a month ago. Any updates on this issue?

**Comment by lauxjpn** [comment]
Came across the same exception while checking the scope of an EF Core bug (unrelated to this issue here) on other providers.

Solved it for me by adding the `SQLitePCLRaw.bundle_e_sqlite3` package reference to the project (I was not using the `Microsoft.EntityFrameworkCore.Sqlite` NuGet package, but a locally compiled build instead):

```xml
<ItemGroup>
    <PackageReference Include="SQLitePCLRaw.core" Version="2.0.4" />
    <PackageReference Include="SQLitePCLRaw.bundle_e_sqlite3" Version="2.0.4" />
</ItemGroup>
```

**Comment by bricelam** [comment]
In that case, this is probably a dupe of https://github.com/ericsink/SQLitePCL.raw/issues/252

**Comment by bairog** [comment]
@bricelam Any updates?

**Comment by bricelam** [comment]
We'd love to get some community help on this issue. Our team has little-to-no experience with C++/CLI. If you or anyone else wants to dig into this, write a detailed explanation of the problem, and propose a fix. We'd be happy do discuss it as a team and possibly move forward with a community PR.

**Comment by bricelam** [comment]
Essentially [this code]() isn't working for some reason on C++/CLI:

https://github.com/dotnet/efcore/blob/786cb40576d47874549753e6d92fbcf3e65a4f64/src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs#L51-L53

Either because SQLitePCLRaw.batteries_v2.dll isn't getting copied to the output directory, or because the type or method is being trimmed.

**Comment by bairog** [comment]
> Either because `SQLitePCLRaw.batteries_v2.dll` isn't getting copied to the output directory, or because the type or method is being trimmed.

As I [wrote above](https://github.com/dotnet/efcore/issues/24493#issuecomment-808695637) `SQLitePCLRaw.batteries_v2.dll` is present in output directory. Don't know if it is trimmed - it is **6Kb**.

**Comment by jjxtra** [comment]
Same issue in .net 6 in vs 2022 debugger, x64 Windows 11 machine, any cpu debug build.

**Comment by bairog** [comment]
@bricelam Any updates? More than a year has passed from bug report date. EF Core 6 has been release and there are some previews of EF Core 7. Did you obtained some help from Microsoft C++\CLI team?

**Comment by bricelam** [comment]
Does manually adding a call to `SQLitePCL.Batteries_V2.Init()` in you app fix the issue?

**Comment by SoftCircuits** [comment]
Is there nothing Microsoft can tell us about this?

I'm using .NET 6.

``` xml
  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="6.0.7">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite.Core" Version="6.0.7" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.7">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>
```

`add-migration` ran without errors.

`update-database` failed.

```
PM> update-database
Build started...
Build succeeded.
System.Exception: You need to call SQLitePCL.raw.SetProvider().  If you are using a bundle package, this is done by calling SQLitePCL.Batteries.Init().
   at SQLitePCL.raw.get_Provider()
   at SQLitePCL.raw.sqlite3_open_v2(utf8z filename, sqlite3& db, Int32 flags, utf8z vfs)
   at SQLitePCL.raw.sqlite3_open_v2(String filename, sqlite3& db, Int32 flags, String vfs)
   at Microsoft.Data.Sqlite.SqliteConnectionInternal..ctor(SqliteConnectionStringBuilder connectionOptions, SqliteConnectionPool pool)
   at Microsoft.Data.Sqlite.SqliteConnectionFactory.GetConnection(SqliteConnection outerConnection)
   at Microsoft.Data.Sqlite.SqliteConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnection(Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternal(Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open(Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Sqlite.Storage.Internal.SqliteDatabaseCreator.Exists()
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.Exists()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
```

Not sure what that was about C++. This is a C# project.

**Comment by bricelam** [comment]
@SoftCircuits I think you meant to use the `Microsoft.EntityFrameworkCore.Sqlite` package instead. The Core one is for those who want to provide their own native SQLite library.
```diff
- <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite.Core" Version="6.0.7" />
+ <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="6.0.7" />
```

**Comment by SoftCircuits** [comment]
@bricelam Sorry, that was the issue. Naming is a little confusing here. I appreciate you taking a look.

**Comment by pajeb4you** [comment]
For my .NET Standard 2.1 project, I was getting same error.  Once I added `Microsoft.Data.Sqlite` package, all started working fine.

**Comment by PaddiM8** [comment]
I'm getting this error with native AOT on .NET 7. I'm using `Microsoft.Data.Sqlite`.

**Comment by bairog** [comment]
@bricelam Same error with latest Microsoft.EntityFrameworkCore.Sqlite **7.0.3** on .**NET 7.0** (latest Visual Studio 2022 **17.5.6** Preview)

> Does manually adding a call to `SQLitePCL.Batteries_V2.Init()` in you app work around the issue?

If I add this before DbContext access - I get an error (despite of **SQLitePCLRaw.batteries_v2.dll** is present in output directory);
> **System.IO.FileNotFoundException:** 'Could not load file or assembly 'SQLitePCLRaw.batteries_v2, Version=2.1.4.1835, Culture=neutral, PublicKeyToken=8226ea5df37bcae9'. Не удается найти указанный файл.'

Did you obtained some help from Microsoft C++\CLI team? This issue exists for almost 2 years :(

**Comment by woutervanranst** [comment]
I had erroneously added this package, which does not include the BCL
<PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite.Core" Version="7.0.4" />

Adding this version instead fixed it for me
<PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="7.0.4" />

**Comment by wz172** [comment]
I encountered the same error, the solution is, package.PackageReference Include = "Microsoft EntityFrameworkCore. Sqlite. Core" Version = "6.0.26" / & gt;Change to &lt;PackageReference Include = "Microsoft. EntityFrameworkCore. Sqlite" Version = "6.0.26" / & gt;

**Comment by gulachek** [comment]
I encountered this in an ASP.NET / ASP.NET Core web app attempting to use Microsoft.Data.Sqlite on Windows 11. It's not the EntityFrameworkCore flavor, but I'm guessing my analysis will be relevant and unlikely to be a separate issue. 

What seems to be going wrong is that ASP.NET "shadow copies" managed assemblies (so C# dlls) to a temp folder. This is where the bubble wrapped SQLitePCL assemblies are loaded from. **The Batteries_V2.Init() function then tries to load e_sqlite3.dll by looking up the native (C/C++) DLL relative to the folder that the SQLitePCL assembly was loaded from. ** Because ASP.NET copies the managed assembly to the temp folder but not the native DLL, the candidate absolute paths that SQLitePCL attempts to load with LoadLibrary are invalid.

These technologies are newer to me so I'm not yet to a point where I can suggest an upstream fix for a PR. I'm happy to provide more investigation details for someone when I'm closer to my dev machine. 

My current status is that I'm trying to find a workaround to get the native DLL to be found by Batteries.Init(). Otherwise I'm going to assess if this bug is enough of a roadblock to try alternative strategies to accomplish my project's goal. If I find a decent workaround, I'll try to remember to update this issue with my findings.

**Comment by chileric** [comment]
> Because ASP.NET copies the managed assembly to the temp folder but not the native DLL, the candidate absolute paths that `SQLitePCL` attempts to load with `LoadLibrary` are invalid.

I'm having this same issue.  @gulachek, were you able to find a solution or a workaround?

**Comment by gulachek** [comment]
> > Because ASP.NET copies the managed assembly to the temp folder but not the native DLL, the candidate absolute paths that `SQLitePCL` attempts to load with `LoadLibrary` are invalid.
> 
> I'm having this same issue.  @gulachek, were you able to find a solution or a workaround?

Hello! Shoot I haven't gotten around to updating this thread yet. The short answer is yes. I need to look back at my project branch to see exactly what I did, but from a high level, I believe that I manually installed the nuget packages that the "batteries included" abstracted away and that enabled me to set up the implementation with a package that worked. I know that's extremely vague and I'll add more details later. This is as much as I remember off-hand from my phone in case the general idea can get you unstuck!

**Comment by chileric** [comment]
Thanks @gulachek.  I kind of get what you're saying and I will try to do the same thing.  I look forward to seeing additional details whenever you have time--that would be awesome!  Thanks so much!

**Comment by chileric** [comment]
> Because ASP.NET copies the managed assembly to the temp folder but not the native DLL, the candidate absolute paths that `SQLitePCL` attempts to load with `LoadLibrary` are invalid.

I finally found a solution that worked for me.  I added this line in my project's Web.config file:

`<hostingEnvironment shadowCopyBinAssemblies="false" />`

Put it inside the `<system.web>` section.  

From what I understand, adding this line made it so my application looked for the missing e_sqlite3.dll file inside my project files inside instead of the Temporary ASP.NET files.  Hope this helps anyone else that is still seeing the dreaded missing e_sqlite3 issue.

**Comment by gulachek** [comment]
### Context

Alright, unfortunately I'm not having an easy time finding the branch I had this on. I also don't think this was impacting the application running on ASP.NET _Core_, and it was only impacting .NET _Framework_. My application doesn't have as strong of needs to support this project on .NET Framework anymore, so it's unlikely I'll invest much into this again, but I'll do my best to detail what I remember doing to work around.

The [Microsoft.Data.Sqlite](https://www.nuget.org/packages/Microsoft.Data.Sqlite/#dependencies-body-tab) package imports both the `Microsoft.Data.Sqlite.Core` and the [SQLitePCLRaw.bundle_e_sqlite3](https://www.nuget.org/packages/SQLitePCLRaw.bundle_e_sqlite3/#dependencies-body-tab) packages. The "sqlite bundle" package, as you can see in the metadata, branches its dependencies differently for NET Framework (4.6.1) and NET Core (NET Standard 2.0). 

In [SQLitePCL.raw/src/common/batteries_v2.cs](https://github.com/ericsink/SQLitePCL.raw/blob/549ecce820170a92b3244ed1205eb81b3f6e50b5/src/common/batteries_v2.cs#L24), you can see that the `Init()` function does a big branching statement depending on how the package is bundled with the native `.dll`, which influences how it sets the provider.

The key workaround I was trying to employ was to _not_ use the "dynamic" version that is defaulted for the .NET Framework bundle (and is subject to the shadow copy issue), and instead use the explicit:

```cs
SQLitePCL.raw.SetProvider(new SQLitePCL.SQLite3Provider_e_sqlite3());
```

### My Hazy Idea of my Workaround

This is where my memory gets fuzzy. I think what I did was, instead of using nuget to install `Microsoft.Data.Sqlite`, I instead took on explicitly installing `Microsoft.Data.Sqlite.Core`, `SQLitePCLRaw.lib.e_sqlite3`, and `SQLitePCLRaw.provider.e_sqlite3`. The last one is the key difference from the default bundle. I then think that I either had to explicitly call `SetProvider` with the `SQLitePCL.SQLite3Provider_e_sqlite3` instance, or the `Batteries.Init()` function did it for me.

I may have needed to reference a native `e_sqlite3` function with `[DllImport]` in one of my `.cs` files to force the runtime to be able to find the native `.dll`.

### What I can Confidently Say

I certainly did get this working in an ASP.NET Framework web app, and you can too!  I'm also confident there are going to be differences in setup between my team's app and the reader's app, and some details are likely going to need to be troubleshooted.

**The key thing to remember is the core functionality in `SQLitePCLRaw` system works, and the issue only lies in setting it up.** The "batteries included" concept makes most cases easier, but it somewhat gets in your way on the legacy web app with the shadow copy behavior.  If you take on the complexity of setting up the system yourself and take some time to inspect the underlying source code, I'm pretty confident you'll be able to get this working (no batteries included 😁).

Regarding the workaround to disable shadow copying, that's great @chileric found something that unblocked you! It might be that simple for some use cases. If you want to review [Shadow Copying Assemblies](https://learn.microsoft.com/en-us/dotnet/framework/app-domains/shadow-copy-assemblies) and make the call for your specific use case to _keep_ the shadow copy behavior, then you'll probably need to do some of the tinkering I described above to get things wired up.

### Final Thoughts

Sorry this wasn't as specific and prescriptive as I was hoping.  I hope this helps any readers who are stuck on this get a bit further!
